include/openrave/environment.h

```diff
@@ -587,6 +587,22 @@ class OPENRAVE_API EnvironmentBase : public boost::enable_shared_from_this<Envir
     /// \throw openrave_exception with ORE_Timeout error code
     virtual void GetBodiesMatchingFilter(std::vector<KinBodyPtr>& bodies, const std::function<bool(const KinBody&)>& filterFunction, uint64_t timeout = 0) const = 0;
 
+    /// \brief Apply a function to every body in the environment. Thread-safe.
+    ///
+    /// This method allows for iterating over all of the bodies in the env without having to copy the list of bodies first.
+    /// The environment interface mutex is locked internally.
+    /// The callback function must not call any methods that would cause bodies to be added or removed from the environment,
+    /// as this would cause a deadlock attempting to exclusively lock the interface mutex while the thread already holds it in shared mode.
+    virtual void IterateBodies(const std::function<void(const KinBodyPtr&)>& mapFunction) = 0;
```

Comment:
Perhaps the function pointer can take in `KinBody&` and allow for modification?

Reply:
You're right, no need to expose the encapsulation here. Changed both instances.

Reply:
Remembered why this contract includes the pointer - it's to allow using this as a way to query a subset of bodies more efficiently, e.g:
```c++
std::vector<KinBodyPtr> bodiesMatchingCriteria;
env->IterateBodies([&bodiesMatchingCriteria](const KinBodyPtr& body) { 
  if (body->MeetsCriteria()) {
    bodiesMatchingCriteria.emplace_back(body);
  }
});
// Outside of interface lock, actually process these bodies...
```
If you have many thousands of bodies, this is a lot more performant than copying _all_ bodies then erasing a bunch of them again. Have reverted the contract here (and explained why in the docstring), but for FilterBodies still only take a kinbody reference.